### PR TITLE
[flutter_tools] prevent NPE due to moved/deleted packages during upgrade packages

### PR DESCRIPTION
**Pull request by jonahwilliams**
## Description

When directories are deleted in git, they can persist locally. Double check that the copy of the SDK we are making excludes these phantom directories when running --update-packages

### CODE REVIEW

**Comment by christopherfujino** [comment]
Why do the directories persist? Is it because gitignored files were generated?

File: packages/flutter_tools/lib/src/commands/update_packages.dart [hunk 1]
```diff
@@ -1425,12 +1425,12 @@ Directory createTemporaryFlutterSdk(
...
 ) {
-  final Set<String> currentPackages = realFlutter
-    .childDirectory('packages')
-    .listSync()
-    .whereType<Directory>()
-    .map((Directory directory) => fileSystem.path.basename(directory.path))
-    .toSet();
+  final Set<String> currentPackages = <String>{};
+  for (final FileSystemEntity entity in realFlutter.childDirectory('packages').listSync()) {
+    if (entity is Directory && entity.childFile('pubspec.yaml').existsSync()) {
```
Comment (christopherfujino) [inline packages/flutter_tools/lib/src/commands/update_packages.dart:1430]: nit could you add a comment?

**Review by christopherfujino — APPROVED** [review]
LGTM with a nit about a comment

**Comment by jonahwilliams** [comment]
git never seems to delete directories

File: packages/flutter_tools/lib/src/commands/update_packages.dart (hunk 1 above)
Comment (jonahwilliams) [reply packages/flutter_tools/lib/src/commands/update_packages.dart:1431]: Done

**Comment by christopherfujino** [comment]
> git never seems to delete directories

i'm fairly sure it will delete a directory on a checkout provided there are no untracked files/diffs.

**Comment by jonahwilliams** [comment]
Ahh, it must be something like .packages or .dart_tool that stick around then?

**Comment by christopherfujino** [comment]
> Ahh, it must be something like .packages or .dart_tool that stick around then?

Yeah. We can land this, it should prevent the NPE. But I wonder if those files hanging around are causing (or will cause) other issues.

**Pull request merged**